### ISSUE TITLE
[BUG] Azure.Data.Tables doesn't support custom type ser/deser

### ISSUE THREAD

**Issue by george-zubrienko** [Azure/azure-sdk-for-net#27413]
### Library name and version

Azure.Data.Tables 12.4.0

### Describe the bug

Given a class:

```
public class CheckpointedRequest : ITableEntity
    {
        [JsonConverter(typeof(UriConverter))]
        public Uri PayloadUri { get; set; }

        [JsonConverter(typeof(UriConverter))]
        public Uri ResultUri { get; set; }

        public string ContentHash { get; set; }
        public string PartitionKey { get; set; }
        public string RowKey { get; set; }
        public DateTimeOffset? Timestamp { get; set; }
        public ETag ETag { get; set; }

}
```

and a converter:

```
        internal class UriConverter : JsonConverter<Uri>
        {
            public override Uri Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
            {
                var value = reader.GetString();
                return string.IsNullOrEmpty(value) switch
                {
                    true => null,
                    false => new Uri(value)
                };
            }

            public override void Write(Utf8JsonWriter writer, Uri value, JsonSerializerOptions options)
            {
                if (value == null)
                {
                    writer.WriteNullValue();
                }
                else
                {
                    writer.WriteStringValue(value.ToString());
                }
            }
        }
```

Calling 
```
tableClient.UpsertEntityAsync(entity, mode: TableUpdateMode.Merge)
```

throws `Not supported type System.Uri` if any of the URI fields are not null.

### Expected behavior

Entity is merged successfully. In the old library I'd do the following:

```
        public override IDictionary<string, EntityProperty> WriteEntity(OperationContext operationContext)
        {
            IDictionary<string, EntityProperty> properties = base.WriteEntity(operationContext);
            properties.Add(nameof(LifeCycleStage), new EntityProperty(LifeCycleStage.ToString()));
            if (PayloadUri != null)
            {
                properties.Add(nameof(PayloadUri), new EntityProperty(PayloadUri.ToString()));
            }

            if (ResultUri != null)
            {
                properties.Add(nameof(ResultUri), new EntityProperty(ResultUri.ToString()));
            }
            
            return properties;
        }
```

and everything works.

### Actual behavior

Error is thrown.

### Reproduction Steps

See description.

### Environment

- .NET 5.0, linux
- VS 16.11

**Comment by george-zubrienko** [comment]
I can also see from exception stack trace that it uses Utf8JsonWriter, but w/o any options. At least we should be able to plug in serialization options?

```
at Azure.Core.Utf8JsonWriterExtensions.WriteObjectValue(Utf8JsonWriter writer, Object value)
 ...
```

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team member best able to assist.

**Comment by christothes** [comment]
Hi @george-zubrienko - Azure.Data.Tables entitites do not support serialization customization via `JsonConverter`. Only property renames and ignoring properties are supported via the `[IgnoreDataMember]` and `[DataMember(Name = "some_new_name")]` attributes.

One possible approach to make this work for your scenario would be to add string properties for the Uri properties which are renamed to the original Uri property names and ignore Uri properties for serialization.

```c#
public class CheckpointedRequest : ITableEntity
{
    [IgnoreDataMember]
    public Uri PayloadUri { get; set; }

    [DataMember(Name = "PayloadUri")]
    public string PayloadUriString { get; set; }

    [IgnoreDataMember]
    public Uri ResultUri { get; set; }

    [DataMember(Name = "ResultUri")]
    public string ResultUriString { get; set; }

    public string ContentHash { get; set; }
    public string PartitionKey { get; set; }
    public string RowKey { get; set; }
    public DateTimeOffset? Timestamp { get; set; }
    public ETag ETag { get; set; }
}

  public class CheckpointedRequest : ITableEntity
{
    private Uri _payloadUri;
    private string _payloadUriString;
    private Uri _resultUri;
    private string _resultUriString;

    [IgnoreDataMember]
    public Uri PayloadUri
    {
        get => _payloadUri;
        set
        {
            _payloadUri = value;
            _payloadUriString = value.ToString();
        }
    }

    [DataMember(Name = "PayloadUri")]
    public string PayloadUriString
    {
        get => _payloadUriString;
        set
        {
            _payloadUri = new Uri(value);
            _payloadUriString = value;
        }
    }

    [IgnoreDataMember]
    public Uri ResultUri
    {
        get => _resultUri;
        set
        {
            _resultUri = value;
            _resultUriString = value.ToString();
        }
    }

    [DataMember(Name = "ResultUri")]
    public string ResultUriString
    {
        get => _resultUriString;
        set
        {
            _resultUri = new Uri(value);
            _resultUriString = value;
        }
    }

    public string ContentHash { get; set; }
    public string PartitionKey { get; set; }
    public string RowKey { get; set; }
    public DateTimeOffset? Timestamp { get; set; }
    public ETag ETag { get; set; }
}
```

**Comment by george-zubrienko** [comment]
@christothes thank you for this workaround. If I can provide some feedback here, this is a significant developer experience downgrade from a previous SDK for Cosmos/ATS. Are there any plans to brink back EntityConverters or maybe expand `ITableEntity` interface, so we can define converters ourselves?

Also I don't think this API currently covers "supports custom entities" category, when it only supports a few types coded in a switch statement, and realistically doesn't support custom types at all.

**Comment by christothes** [comment]
@george-zubrienko  Thanks for the feedback.

The custom type support for Tables is restricted by the types supported by the service which are documented [here](https://docs.microsoft.com/en-us/rest/api/storageservices/understanding-the-table-service-data-model#property-types). Although the previous client library was more flexible with conversions, ultimately these restrictions were always in place.

If you have any specific blocking scenarios with the new Azure.Data.Tables library's approach to serialization, please let us know and we'll do our best to assist.

**Comment by george-zubrienko** [comment]
> @george-zubrienko  Thanks for the feedback.
> 
> 
> 
> The custom type support for Tables is restricted by the types supported by the service which are documented [here](https://docs.microsoft.com/en-us/rest/api/storageservices/understanding-the-table-service-data-model#property-types). Although the previous client library was more flexible with conversions, ultimately these restrictions were always in place.
> 
> 
> 
> If you have any specific blocking scenarios with the new Azure.Data.Tables library's approach to serialization, please let us know and we'll do our best to assist. 

I am aware of basic ATS type restrictions. But legacy lib was really good at providing a way to store any objects there, as long as a developer provides a serializer. Just look at the workaround you suggested for a new lib, it's a code smell we now have to add, because you guys decided nobody needs to store anything but 3 basic types w/o any conversion support.
It's not a blocker, it is just a very sad direction for once a decent SDK.
Thanks for clarifications though, closing the issue.

**Comment by george-zubrienko** [comment]
Just a follow-up for those who might encounter the similar issue. Instead of using the suggested approach, we went with throwing away inheritance from `ITableEntity` and instead adding converter methods from and to `TableEntity` sealed class.

```
    public class StreamInfo
    {
        public string Id { get; set; }

        [JsonConverter(typeof(JsonStringEnumConverter))]
        public StreamSources StreamSource { get; set; }

        public DateTimeOffset? StartedAt { get; set; }

        public TableEntity ToTableEntity()
        {
            return new TableEntity(StreamSource.ToString(), Id)
            {
                { nameof(StreamSource), StreamSource },
                { nameof(StartedAt), StartedAt }
            };
        }

        /// <summary>
        /// Reads this stream back from table storage.
        /// </summary>
        /// <param name="te"></param>
        /// <returns></returns>
        public static StreamInfo FromTableEntity(TableEntity te)
        {
            return new StreamInfo
            {
                Id = te.RowKey,
                StreamSource = (StreamSources)Enum.Parse(typeof(StreamSources), te.PartitionKey),
                StartedAt = te.GetDateTimeOffset(nameof(StartedAt))
            };
        }

        public static StreamInfo FromTableEntity(object te)
        {
            if (te.GetType() == typeof(TableEntity))
            {
                return FromTableEntity(te as TableEntity);
            }
            else
            {
                return null;
            }
        }
```

**Comment by chkeita** [comment]
> If you have any specific blocking scenarios with the new Azure.Data.Tables library's approach to serialization, please let us know and we'll do our best to assist.

### Case conversion

@christothes We are porting an existing service that was originally written in python. The column names are all `snake_case` following the python convention. I would like to have the entity created in C# to follow the `PascalCase` convention. 
Unfortunately, We don't  have the option of renaming the column because there are other services that depends on the DB. 

so for each entity we have to annotate the field like this.

``` C#
   record MyEntity: ITableEntity
    {
	    [DataMember(Name ="property_1")]
	    public Propery1: string
	    [DataMember(Name ="property_2")]
	    public Propery2: string
    }
```
basically, all fields need to be annotated with  `[DataMember(Name =<fieldName.ToSnakeCase()> )]`
It would be nice and less error prone to be able to do this in one place. With the support of a converter this should be possible.

**Comment by dolenox** [comment]
This looks like it was an issue before and there was an additional to the previous SDK to add utilities to help:

https://github.com/Azure/azure-storage-net/pull/337/commits/daff940a506b27650901b9b9feb59131ffacce6d#diff-cb30fec64299b63bf4b6521e5b7f22044cede09cb3130f93179866bd7b917d05

From the third comment here:
https://stackoverflow.com/questions/19885219/insert-complex-objects-to-azure-table-with-tableserviceentity

Was this functionality removed in newer versions?

**Comment by Yashuaa** [comment]
> > @george-zubrienko  Thanks for the feedback.
> > The custom type support for Tables is restricted by the types supported by the service which are documented [here](https://docs.microsoft.com/en-us/rest/api/storageservices/understanding-the-table-service-data-model#property-types). Although the previous client library was more flexible with conversions, ultimately these restrictions were always in place.
> > If you have any specific blocking scenarios with the new Azure.Data.Tables library's approach to serialization, please let us know and we'll do our best to assist.
> 
> I am aware of basic ATS type restrictions. But legacy lib was really good at providing a way to store any objects there, as long as a developer provides a serializer. Just look at the workaround you suggested for a new lib, it's a code smell we now have to add, because you guys decided nobody needs to store anything but 3 basic types w/o any conversion support. It's not a blocker, it is just a very sad direction for once a decent SDK. Thanks for clarifications though, closing the issue.

@george-zubrienko  is correct. This was a crappy move by MSFT, and due to use upgrading this nuget it broke live production code.

These are breaking changes Microsoft has made... The issue is that "Azure Storage Table no longer supports custom types unlike the last version which _did_"
